File: mobile/android/core/network/src/main/kotlin/com/qervon/core/network/Requests.kt
// =============================================================================
// File:           mobile/android/core/network/src/main/kotlin/com/qervon/core/network/Requests.kt
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Request/response wire DTOs that only exist at the HTTP boundary
//   (mirrors the private structs declared inline in
//   `backend/apps/api-gateway/src/http.rs`, e.g. `LoginRequest`,
//   `AuthResponse`, `OtpRequestRequest`).
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

package com.qervon.core.network

import com.qervon.core.common.model.Address
import kotlinx.serialization.Serializable

@Serializable
data class LoginBody(val email: String, val password: String, val tenantSlug: String)

@Serializable
data class RegisterAccountBody(
    val email: String,
    val displayName: String,
    val password: String,
    val tenantSlug: String? = null,
)

@Serializable
data class RefreshBody(val refreshToken: String)

@Serializable
data class OtpRequestBody(val tenantSlug: String, val phone: String)

@Serializable
data class OtpRequestResponseBody(val status: String, val devCode: String? = null)

@Serializable
data class OtpVerifyBody(val tenantSlug: String, val phone: String, val code: String)

@Serializable
data class SetPhoneBody(val phone: String)

@Serializable
data class AuthResponseBody(
    val accessToken: String,
    val refreshToken: String,
    val tokenType: String,
    val expiresInSeconds: Long,
)

@Serializable
data class SetCourierAvailabilityBody(val online: Boolean)

@Serializable
data class UpdateLocationBody(
    val latitude: Double,
    val longitude: Double,
    val speedKmh: Double? = null,
    val batteryPct: Int? = null,
)

@Serializable
data class CompleteDeliveryBody(
    val recipientName: String,
    val qrBarcodeVerified: Boolean,
    val digitalSignatureBase64: String? = null,
    val photoEvidenceUrl: String? = null,
    val paymentCollected: Boolean = false,
)

@Serializable
data class UploadedFileResponseBody(val url: String)

@Serializable
data class CreateCustomerOrderBody(
    val pickup: Address,
    val dropoff: Address,
    val couponCode: String? = null,
    val paymentMethod: String? = null,
    val deliveryNote: String? = null,
    val contactPhone: String? = null,
)

@Serializable
data class CreateCustomerAddressBody(
    val label: String,
    val latitude: Double,
    val longitude: Double,
    val fullAddress: String,
)

@Serializable
data class RateOrderBody(val ratingStars: Int, val comment: String? = null)

@Serializable
data class OpenSupportTicketBody(
    val orderId: String? = null,
    val subject: String,
    val message: String,
)

File: mobile/ios/Packages/QervonKit/Sources/QervonNetworking/Requests.swift
// =============================================================================
// File:           mobile/ios/Packages/QervonKit/Sources/QervonNetworking/Requests.swift
// Project:        Qervon
// Author:         USDTG GROUP TECHNOLOGY LLC
// Developer:      Irfan Gedik
// Created Date:   2026-08-12
// Version:        0.1.0
//
// Description:
//   Request/response body shapes specific to individual endpoints (as
//   opposed to the shared domain models in `QervonCore`).
//
// License:
//   Qervon License v1.0 — see LICENSE in the repository root.
// =============================================================================

import Foundation
import QervonCore

struct OtpRequestBody: Encodable {
    let tenantSlug: String
    let phone: String
}

public struct OtpRequestResult: Decodable, Sendable {
    public let status: String
    /// Only populated on in-memory (local/dev) backend storage — see
    /// BACKEND_BACKLOG.md. Never populated against a real deployment.
    public let devCode: String?
}

struct OtpVerifyBody: Encodable {
    let tenantSlug: String
    let phone: String
    let code: String
}

struct LoginBody: Encodable {
    let email: String
    let password: String
    let tenantSlug: String
}

struct RefreshBody: Encodable {
    let refreshToken: String
}

struct LinkPhoneBody: Encodable {
    let phone: String
}

struct SetAvailabilityBody: Encodable {
    let online: Bool
}

public struct UpdateLocationBody: Encodable, Sendable {
    public let latitude: Double
    public let longitude: Double
    public let speedKmh: Double?
    public let batteryPct: Double?

    public init(latitude: Double, longitude: Double, speedKmh: Double?, batteryPct: Double?) {
        self.latitude = latitude
        self.longitude = longitude
        self.speedKmh = speedKmh
        self.batteryPct = batteryPct
    }
}

public struct DeliverOrderBody: Encodable, Sendable {
    public let recipientName: String
    public let qrBarcodeVerified: Bool
    public let digitalSignatureBase64: String?
    public let photoEvidenceUrl: String?
    public let paymentCollected: Bool

    public init(
        recipientName: String,
        qrBarcodeVerified: Bool,
        digitalSignatureBase64: String?,
        photoEvidenceUrl: String?,
        paymentCollected: Bool
    ) {
        self.recipientName = recipientName
        self.qrBarcodeVerified = qrBarcodeVerified
        self.digitalSignatureBase64 = digitalSignatureBase64
        self.photoEvidenceUrl = photoEvidenceUrl
        self.paymentCollected = paymentCollected
    }
}

public struct CompletePickupBody: Encodable, Sendable {
    public let pickupPhotoEvidenceUrl: String

    public init(pickupPhotoEvidenceUrl: String) {
        self.pickupPhotoEvidenceUrl = pickupPhotoEvidenceUrl
    }
}

struct RegisterPushDeviceBody: Encodable {
    let platform: String
    /// "courier" or "customer" — see `AppVariant`. Tells the backend which
    /// bundle id's `apns-topic` this token belongs to.
    let app: String
    let deviceToken: String
}

struct UploadedFileResponse: Decodable {
    let url: String
}

/// Empty JSON object body (`{}`), used for endpoints that ignore any body.
struct EmptyBody: Encodable {}

// MARK: - Customer

struct RegisterAccountBody: Encodable {
    let email: String
    let displayName: String
    let password: String
    let tenantSlug: String?
}

struct AddAddressBody: Encodable {
    let label: String
    let latitude: Double
    let longitude: Double
    let fullAddress: String
}

public struct CreateCustomerOrderBody: Encodable, Sendable {
    public let pickup: Address
    public let dropoff: Address
    public let couponCode: String?
    public let paymentMethod: String?
    public let deliveryNote: String?
    public let contactPhone: String?

    public init(
        pickup: Address,
        dropoff: Address,
        couponCode: String?,
        paymentMethod: PaymentMethod?,
        deliveryNote: String?,
        contactPhone: String?
    ) {
        self.pickup = pickup
        self.dropoff = dropoff
        self.couponCode = couponCode
        self.paymentMethod = paymentMethod?.rawValue
        self.deliveryNote = deliveryNote
        self.contactPhone = contactPhone
    }
}

struct RateOrderBody: Encodable {
    let ratingStars: Int
    let comment: String?
}

struct OpenSupportTicketBody: Encodable {
    let orderId: UUID?
    let subject: String
    let message: String
}
